README.md
# lkapi

[![Awesome](https://awesome.re/badge.svg)](https://michaelpeterswa.com) [![start with why](https://img.shields.io/badge/start%20with-why%3F-brightgreen.svg?style=flat)](http://www.ted.com/talks/simon_sinek_how_great_leaders_inspire_action)
[![codecov](https://codecov.io/gh/michaelpeterswa/lkapi/branch/master/graph/badge.svg)](https://codecov.io/gh/michaelpeterswa/lkapi) [![Known Vulnerabilities](https://snyk.io/test/github/michaelpeterswa/lkapi/badge.svg)](https://snyk.io/test/github/michaelpeterswa/lkapi)  ![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg) ![GitHub repo size](https://img.shields.io/github/repo-size/michaelpeterswa/lkapi.svg) ![GitHub issues](https://img.shields.io/github/issues/michaelpeterswa/lkapi.svg) [![Maintainability](https://api.codeclimate.com/v1/badges/3876447b891bfba74e1c/maintainability)](https://codeclimate.com/github/michaelpeterswa/lkapi/maintainability)

this is a REST api interface to go alongside LetterkennyQuotes
```
curl -s "https://cascades.dev/api" | jq -r '.quote' | lolcat -f
```
> this api uses Express to serve a json file including a quote, number, and designation

![Alt text](img/lkapi1.png?raw=true "Webpage")
![Alt text](img/lkapi2.png?raw=true "JSON endpoint")

## Installation
```
git clone https://github.com/michaelpeterswa/lkapi.git
```

Or download the file manually.
## Release History
* 0.0.1
   * Opened Repository (07.09.2019)
## Meta
Michael Peters - michael@michaelpeterswa.com

Distributed under the MIT license. See ``LICENSE`` for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 0.1.0 (2020-12-16)


### ⚠ BREAKING CHANGES

* change endpoints and file support, condense to one file

### Features

* change endpoints and file support, condense to one file ([9cb7105](https://github.com/michaelpeterswa/lkapi/commit/9cb710573deafe4b2527f92034242c060a62a181))


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([8c4ebe8](https://github.com/michaelpeterswa/lkapi/commit/8c4ebe85e2cd998a763f01cb0b19e3958f32e111))
* package.json & package-lock.json to reduce vulnerabilities ([ba046a8](https://github.com/michaelpeterswa/lkapi/commit/ba046a846060a08173ea968e4e7f4ba9afc8f53e))
* package.json & package-lock.json to reduce vulnerabilities ([e0b85f1](https://github.com/michaelpeterswa/lkapi/commit/e0b85f1d6dd1650e65d1b1b121e2d787440bb07f))


test.sh
#!/bin/bash

echo "testing the codebase..."
npx mocha --exit

test/test.js
/* eslint-disable no-undef */
// letterkennyQuotes api
// TESTING FILE
// Michael Peters

var request = require('supertest');
var app = require('../src/server.js');

describe('GET /', function() {
    it('main webpage responds with html and status 200', function(done) {
    request(app)
        .get('/')
        .expect('Content-Type', /html/)
        .expect(200, done);
    });
});

describe('GET /api', function() {
    it('main endpoint responds with valid json api output', function(done) {
    request(app)
        .get('/api')
        .expect('Content-Type', /json/)
        .expect(200, done);
    });
});

describe('GET /sfw', function() {
    it('main endpoint responds with valid sfw json api output', function(done) {
    request(app)
        .get('/sfw')
        .expect('Content-Type', /json/)
        .expect(200, done);
    });
});

describe('GET /nsfw', function() {
    it('main endpoint responds with valid nsfw json api output', function(done) {
    request(app)
        .get('/nsfw')
        .expect('Content-Type', /json/)
        .expect(200, done);
    });
});


describe('GET /test', function() {
    it('tests an invalid link and ensures status 404', function(done) {
    request(app)
        .get('/test')
        .expect('Content-Type', /html/)
        .expect(404, done);
    });
});

describe('GET /styles/styles.css', function() {
    it('tests for presence of stylesheet', function(done) {
    request(app)
        .get('/styles/styles.css')
        .expect('Content-Type', /css/)
        .expect(200, done);
    });
});

test/test.sh
#!/bin/bash

echo "testing the codebase..."
npx nyc --reporter=json-summary --reporter=html --reporter=cobertura --reporter=text-summary mocha --exit

src/server.js
// letterkennyQuotes api
// Michael Peters
// July 12, 2019
// revived in time for Season 9 -- December 16, 2020

const express = require('express')
var cors = require('cors')
var fs = require('fs')
var path = require('path');

const app = express()

const mainPath = "./resources/main.json"

//port the app is currently serving to
const port = 6969

var content
var quote
var sfw
var id = 0

// eslint-disable-next-line no-unused-vars
function readFileWithPathAndContent(path, callback) {
    fs.readFile(path, 'utf8', function read(err, data) {
        if (err) {
            data = "[\"an error occured - file not found\"]"
        }
        return callback(data)
    })
    
}

/// Read from the main.json file //
readFileWithPathAndContent(mainPath, function(response){
    content = response
    })
////////////////////////////////

function generateRandomWithLength(length){
    var randomNum = Math.floor(Math.random() * length)
    return randomNum
}

function setQuoteAndData(num, quoteData){
        quote = quoteData[0]
        id = num
        sfw = quoteData[1]    
}

function processFile(message, status) {
    var data = JSON.parse(message)
    var length = data.all.length
    var randomNum = generateRandomWithLength(length)
    data = data.all
    if (status === "all"){
        var randomQuote = data[randomNum]
        setQuoteAndData(randomNum, randomQuote)
    }
    else if (status === "sfw"){
        randomQuote = data[randomNum]
        while(randomQuote[1] != "sfw"){
            randomNum = generateRandomWithLength(length)
            randomQuote = data[randomNum]
        }
        setQuoteAndData(randomNum, randomQuote)      
    }
    else {
        randomQuote = data[randomNum]
        while(randomQuote[1] != "nsfw"){
            randomNum = generateRandomWithLength(length)
            randomQuote = data[randomNum]
        }
        setQuoteAndData(randomNum, randomQuote)
    }
}

app.use(cors());

app.get('/', function(req, res) {
    res.sendFile(path.join(__dirname + '/../html/index.html'));
  });

//lkapi main endpoint (JSON)
app.get('/api', function (req, res) {
    processFile(content, "all")
    res.status(200).json({
    quote: quote,
    id: id,
    sfw: sfw
    })
})
app.get('/sfw', function (req, res) {
    processFile(content, "sfw")
    res.status(200).json({
    quote: quote,
    id: id,
    sfw: sfw
    })
})
app.get('/nsfw', function (req, res) {
    processFile(content, "nsfw")
    res.status(200).json({
    quote: quote,
    id: id,
    sfw: sfw
    })
})

app.get('/styles/styles.css', function(req, res) {
    res.sendFile(path.join(__dirname + '/../html/styles/styles.css'));
  });

app.use(function (req, res) {
    res.status(404).sendFile(path.join(__dirname + '/../html/404.html'));
})

const server = app.listen(port, () => console.log(`LetterkennyAPI server app listening on port ${port}!\n`))

module.exports = server